linkerd/app/src/identity.rs

```diff
@@ -55,47 +62,85 @@ impl Config {
         client_metrics: ClientMetrics,
         registry: &mut prom::Registry,
     ) -> Result<Identity> {
-        let name = self.params.server_name.clone();
-        let (store, receiver) = Mode::default().watch(
-            name.clone().into(),
-            name.clone(),
-            &self.params.trust_anchors_pem,
-        )?;
-
-        let certify = Certify::from(self.certify);
-
-        let addr = self.control.addr.clone();
-
-        let (tx, ready) = watch::channel(false);
-
-        // Save to be spawned on an auxiliary runtime.
-        let task = Box::pin({
-            let addr = addr.clone();
-            let svc = self.control.build(
-                dns,
-                client_metrics,
-                registry.sub_registry_with_prefix("control_identity"),
-                receiver.new_client(),
-            );
-
-            let cert_metrics =
-                CertMetrics::register(registry.sub_registry_with_prefix("identity_cert"));
-            let cred = WithCertMetrics::new(cert_metrics, NotifyReady { store, tx });
-
-            certify
-                .run(name, cred, svc)
-                .instrument(tracing::debug_span!("identity", server.addr = %addr).or_current())
-        });
-
-        Ok(Identity {
-            addr,
-            receiver,
-            ready,
-            task,
+        let cert_metrics =
+            CertMetrics::register(registry.sub_registry_with_prefix("identity_cert"));
+
+        Ok(match self {
+            Self::Linkerd {
+                client,
+                certify,
+                tls,
+            } => {
+                // TODO: move this validation into env.rs
+                let name = match (&tls.id, &tls.server_name) {
+                    (Id::Dns(id), sni) if id == sni => id.clone(),
+                    (_id, _sni) => {
+                        return Err(
+                            "Linkerd identity requires a TLS Id and server name to be the same"
```

Comment:
Ideally we'd use thiserror again here. Since is up at the app layer I don't care as much as I would in a lib.